File: examples/go/lmstream/frame.go
package lmstream

import (
	"encoding/binary"
	"fmt"
)

const (
	HeaderSize     = 8
	MaxPayloadSize = 65527
)

type LMStreamError struct {
	Code    ErrorCode
	Scope   ErrorScope
	Message string
}

func (e *LMStreamError) Error() string {
	scopeStr := "SESSION"
	if e.Scope == ErrorScopeConnection {
		scopeStr = "CONNECTION"
	}
	return fmt.Sprintf("[LMStream %s ERROR %d] %s", scopeStr, e.Code, e.Message)
}

type FrameHeader struct {
	StreamID uint32
	Opcode   Opcode
	Flags    uint8
	Length   uint16
}

type Frame struct {
	Header  FrameHeader
	Payload []byte
}

func EncodeHeader(h FrameHeader) ([]byte, error) {
	if int(h.Length) > MaxPayloadSize {
		return nil, &LMStreamError{Code: ErrorCodeProtocolViolation, Scope: ErrorScopeConnection, Message: "Length exceeds cap"}
	}
	buf := make([]byte, HeaderSize)
	binary.LittleEndian.PutUint32(buf[0:4], h.StreamID)
	buf[4] = uint8(h.Opcode)
	buf[5] = h.Flags
	binary.LittleEndian.PutUint16(buf[6:8], h.Length)
	return buf, nil
}

func DecodeHeader(b []byte) (FrameHeader, error) {
	if len(b) < HeaderSize {
		return FrameHeader{}, &LMStreamError{Code: ErrorCodeProtocolViolation, Scope: ErrorScopeConnection, Message: "Header too short"}
	}
	streamID := binary.LittleEndian.Uint32(b[0:4])
	opRaw := b[4]
	flags := b[5]
	length := binary.LittleEndian.Uint16(b[6:8])

	if int(length) > MaxPayloadSize {
		return FrameHeader{}, &LMStreamError{Code: ErrorCodeProtocolViolation, Scope: ErrorScopeConnection, Message: "Length exceeds cap"}
	}

	if (flags & HeaderFlagReservedMask) != 0 {
		return FrameHeader{}, &LMStreamError{Code: ErrorCodeProtocolViolation, Scope: ErrorScopeConnection, Message: "Reserved flags set"}
	}

	op := Opcode(opRaw)
	isKnownCore := opRaw >= 0x01 && opRaw <= 0x11
	if !isKnownCore && opRaw <= 0x2F {
		return FrameHeader{}, &LMStreamError{Code: ErrorCodeProtocolViolation, Scope: ErrorScopeConnection, Message: "Unknown core opcode"}
	}

	isConn := op == OpcodeSettings || op == OpcodePing || op == OpcodePong || op == OpcodeError
	if streamID == 0 && !isConn {
		return FrameHeader{}, &LMStreamError{Code: ErrorCodeProtocolViolation, Scope: ErrorScopeConnection, Message: "Data opcode on StreamID 0"}
	}
	if streamID != 0 && (op == OpcodeSettings || op == OpcodePing || op == OpcodePong) {
		return FrameHeader{}, &LMStreamError{Code: ErrorCodeProtocolViolation, Scope: ErrorScopeConnection, Message: "Control opcode on non-zero stream"}
	}

	// Fixed length validation
	fixedLengths := map[Opcode]int{
		OpcodeSettings: 18,
		OpcodePing:     8,
		OpcodePong:     8,
		OpcodeUsage:    20,
		OpcodeMetrics:  20,
		OpcodeEnd:      2,
		OpcodeCancel:   0,
		OpcodePause:    0,
		OpcodeResume:   0,
		OpcodeRewind:   8,
	}
	if reqLen, ok := fixedLengths[op]; ok && int(length) != reqLen {
		return FrameHeader{}, &LMStreamError{Code: ErrorCodeProtocolViolation, Scope: ErrorScopeConnection, Message: "Fixed length mismatch"}
	}

	return FrameHeader{StreamID: streamID, Opcode: op, Flags: flags, Length: length}, nil
}

func EncodeFrame(f Frame) ([]byte, error) {
	f.Header.Length = uint16(len(f.Payload))
	hBytes, err := EncodeHeader(f.Header)
	if err != nil {
		return nil, err
	}
	return append(hBytes, f.Payload...), nil
}

func DecodeFrame(b []byte) (Frame, error) {
	h, err := DecodeHeader(b)
	if err != nil {
		return Frame{}, err
	}
	total := HeaderSize + int(h.Length)
	if len(b) < total {
		return Frame{}, &LMStreamError{Code: ErrorCodeProtocolViolation, Scope: ErrorScopeConnection, Message: "Truncated payload"}
	}
	return Frame{Header: h, Payload: b[HeaderSize:total]}, nil
}

File: examples/python/lmstream/frame.py
import struct
from dataclasses import dataclass
from typing import Tuple
from .enums import Opcode, HeaderFlags, ErrorCode

HEADER_SIZE = 8
MAX_PAYLOAD_SIZE = 65527

class LMStreamError(Exception):
    def __init__(self, code: ErrorCode, scope: str, message: str):
        super().__init__(f"[LMStream {scope.upper()} ERROR {code}] {message}")
        self.code = code
        self.scope = scope
        self.message = message

@dataclass
class FrameHeader:
    stream_id: int
    opcode: Opcode
    flags: int
    length: int

@dataclass
class Frame:
    header: FrameHeader
    payload: bytes

def encode_header(h: FrameHeader) -> bytes:
    if h.length > MAX_PAYLOAD_SIZE:
        raise LMStreamError(ErrorCode.PROTOCOL_VIOLATION, "connection", f"Payload length {h.length} exceeds cap {MAX_PAYLOAD_SIZE}")
    return struct.pack("<IBBH", h.stream_id, int(h.opcode), h.flags, h.length)

def decode_header(b: bytes) -> FrameHeader:
    if len(b) < HEADER_SIZE:
        raise LMStreamError(ErrorCode.PROTOCOL_VIOLATION, "connection", f"Buffer too short: {len(b)}")
    stream_id, op_raw, flags, length = struct.unpack("<IBBH", b[:HEADER_SIZE])

    if length > MAX_PAYLOAD_SIZE:
        raise LMStreamError(ErrorCode.PROTOCOL_VIOLATION, "connection", f"Length {length} > {MAX_PAYLOAD_SIZE}")

    if (flags & HeaderFlags.RESERVED_MASK) != 0:
        raise LMStreamError(ErrorCode.PROTOCOL_VIOLATION, "connection", f"Reserved flags set: 0x{flags:x}")

    try:
        opcode = Opcode(op_raw)
    except ValueError:
        if 0x00 <= op_raw <= 0x2F:
            raise LMStreamError(ErrorCode.PROTOCOL_VIOLATION, "connection", f"Unknown core opcode 0x{op_raw:x}")
        opcode = op_raw

    is_conn = opcode in (Opcode.SETTINGS, Opcode.PING, Opcode.PONG, Opcode.ERROR)
    if stream_id == 0 and not is_conn:
        raise LMStreamError(ErrorCode.PROTOCOL_VIOLATION, "connection", f"Data opcode 0x{op_raw:x} on StreamID 0")

    if stream_id != 0 and opcode in (Opcode.SETTINGS, Opcode.PING, Opcode.PONG):
        raise LMStreamError(ErrorCode.PROTOCOL_VIOLATION, "connection", f"Control opcode on non-zero StreamID {stream_id}")

    # Validate fixed-size opcode payload lengths
    fixed_lengths = {
        Opcode.SETTINGS: 18,
        Opcode.PING: 8,
        Opcode.PONG: 8,
        Opcode.USAGE: 20,
        Opcode.METRICS: 20,
        Opcode.END: 2,
        Opcode.CANCEL: 0,
        Opcode.PAUSE: 0,
        Opcode.RESUME: 0,
        Opcode.REWIND: 8,
    }
    if opcode in fixed_lengths and length != fixed_lengths[opcode]:
        raise LMStreamError(
            ErrorCode.PROTOCOL_VIOLATION,
            "connection",
            f"Opcode 0x{op_raw:x} requires exact payload length {fixed_lengths[opcode]}, got {length}"
        )

    return FrameHeader(stream_id, opcode, flags, length)

def encode_frame(f: Frame) -> bytes:
    h = f.header
    h.length = len(f.payload)
    return encode_header(h) + f.payload

def decode_frame(b: bytes) -> Frame:
    h = decode_header(b)
    total = HEADER_SIZE + h.length
    if len(b) < total:
        raise LMStreamError(ErrorCode.PROTOCOL_VIOLATION, "connection", f"Truncated frame: expected {h.length}, got {len(b)-HEADER_SIZE}")
    return Frame(h, b[HEADER_SIZE:total])
